### PR TITLE
More complex quick start examples to introduce other concepts

### PR DESCRIPTION
**Pull request by lanthias**
## What does this PR do?
I've added another section in quick_start to:
 - Build a more complex module (a classifier that uses some system libraries)
 - Introduced new sources and sinks
 - Introduce partial workflows
 - Introduce workflow modules

## Notes
 - It doesn't yet include more complex types
 - It is badly organised and needs to be split into separate files

### CODE REVIEW

File: quick_start/more.rst [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+
+Now we're ready to build and publish our classifier. Remember, even though we run this command locally, our module gets built and published to your NStack Server.
+
+.. code :: bash
+
+    ~/irisclassify/ $ nstack build
+    Building NStack Container module irisclassify. Please wait. This may take some time.
+    Module irisclassify built successfully. Use `nstack list methods` to see all available methods.
+
+We can see our method, ``irisclassify.predict``. Including our ``demo.numChars`` method from the previous tutorial, we should now have two:
```
Comment (NickPollard) [inline quick_start/more.rst]: This line reads oddly - presumably you mean something like:

> We can now see ``irisclassify.predict`` in the list of existing methods (along with previously built methods like ``demo.numChars``) by running the suggested command `nstack list methods`

File: quick_start/more.rst [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+.. code :: bash
+   
+    ~/irisclassify/ $ nstack start "source(http:///irisendpoint : (Double, Double, Double, Double)) | irisclassify.predict | sink(log:// : Text)"
+
+This creates an HTTP endpoint on ``http://localhost:8080/irisendpoint`` which can receive four ``Double``\s, and writes the results to the log as ``Text``. We can test our classifier by sending it some of the sample data from ``train.csv``:
+
+.. code :: bash
+
+   ~/irisclassify/ $ curl -X PUT -d '{ "params" : [4.7, 1.4, 6.1, 2.9] }' localhost:8080/irisendpoint 
+   Success
```
Comment (NickPollard) [inline quick_start/more.rst]: Instead of `Success` we now see `Msg Accepted` here.

File: quick_start/more.rst [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+   ~/irisclassify/ $ nstack log 2  
+   Feb 17 10:32:30 nostromo nstack-server[8925]: OUTPUT: "Iris-versicolor"
+
+Great! Our classifiier is now productionised.
```
Comment (NickPollard) [inline quick_start/more.rst]: typo in `classifiier`

File: quick_start/more.rst [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+Other Sources and Sinks
```
Comment (NickPollard) [inline quick_start/more.rst]: I wonder if a better header here would be `Connecting to a Database` - it's more user-story focused, perhaps more compelling?

File: quick_start/more.rst (hunk 1 above)
Comment (lanthias) [reply quick_start/more.rst]: Yes, thank you

File: quick_start/more.rst (hunk 2 above)
Comment (lanthias) [reply quick_start/more.rst]: Thank you

File: quick_start/more.rst [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+    def completeWorkflow = source(http:///speciesLength : (Double, Double, Double, Double)) | irisworkflow.speciesLength | sink(log:// : Integer)
+  }
+
+If you ``build`` this, you can then start it by itself with the ``start`` command, because it's a fully composed:
+
+.. code :: bash
+
+  ~/irisworkflow/ $ nstack start irisworkflow.completeWorkflow
+
+This paradigm can be helpful when we apply it to sources and sinks. Oftentimes, you -- or someone else in your company -- will want to create sources and sinks which are combined with modules, for instance in the following fictional example:
```
Comment (NickPollard) [inline quick_start/more.rst]: Maybe we should build a practical example here that someone can go through, for defining sources and sinks?

E.g. define a reusable `http` endpoint:

```
def irisEndpoint = source(http:///irisclassify : (Double,Double,Double,Double))
```

Which they can then use and test? Or we stick with Postgres but make sure we actually have a test db on the server they can use.

Comment (lanthias) [reply quick_start/more.rst]: I like the former, I think we should mention the latter as well though.

**Comment by lanthias** [comment]
The last more complex example should probably be split into sections IMO

File: quick_start/more.rst [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    from sklearn.ensemble import RandomForestClassifier
+
+.. note :: Python modules must also import ``nstack``
+
+Before we add our ``predict`` method, we're going to add ``__init__``, the Python contructor method which runs upon the creation of our module. It's going to load our data from ``train.csv``, and use it to train our Random Forest classifier:
+
+.. code :: python
+
+    def __init__(self):
+        train = pd.read_csv("train.csv")
```
Comment (rjmk) [inline quick_start/more.rst]: Even our examples escape our type system!

Not a criticism as it seems like its actually the main use case

File: advanced_start/index.rst [hunk 7]
```diff
@@ -0,0 +1,15 @@
+.. _advanced_start_index:
+
+******************
+Advanced Tutorial
+******************
+
+In this section, we're going to productionise a Random Forest classifier written with `sklearn`, deploy it to the cloud, and use it in a more sophisticated workflow.
```
Comment (jonboulle) [inline advanced_start/index.rst]: link to what sklearn is (and probably likewise with random forest)

File: advanced_start/more.rst [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+To begin, let's make a new directory called ``irisclassify``, ``cd`` into it, and initialise a new module:
+
+.. code:: bash
+    
+    ~/ $ mkdir irisclassify; cd irisclassify
+    ~/irisclassify/ $ nstack init python
+    python module 'irisclassify' successfully initialised at ~/irisclassify
+
+Next, let's download our training data into this so we can use it in our module.
```
Comment (jonboulle) [inline advanced_start/more.rst]: into this -> into this directory

Comment (jonboulle) [reply advanced_start/more.rst]: (or just remove it)

File: advanced_start/more.rst [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+.. code:: bash
+
+    ~/irisclassify/ $ curl -O https://raw.githubusercontent.com/nstackcom/nstack-examples/master/iris/irisclassify/train.csv
+
+
+Defining our API
+****************
+
+As we know what the input and output of our classifier is going to look like, let's edit the ``api`` section of ``nstack.yaml`` to define our API (i.e. the entry-point into our module). By default a new module contains a sample function ``numChars``, which we can replace with our definition. We're going to call the function we write in Python ``predict``, which means we can fill in the ``api`` section of ``nstack.yaml`` as follows:
```
Comment (jonboulle) [inline advanced_start/more.rst]: we can fill in -> we fill in

File: advanced_start/more.rst [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+.. code :: python
+
+    import nstack
+    import pandas as pd
+
+    from sklearn.ensemble import RandomForestClassifier
+
+.. note :: Python modules must also import ``nstack``
+
+Before we add our ``predict`` function, we're going to add ``__init__``, the Python contructor function which runs upon the creation of our module. It's going to load our data from ``train.csv``, and use it to train our Random Forest classifier:
```
Comment (jonboulle) [inline advanced_start/more.rst]: contructor -> constructor

File: advanced_start/more.rst [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        return results.item()
+
+Configuring your module
+***********************
+
+When your module is started, it is run in a Linux container on the NStack server. Because our module uses libraries like ``pandas`` and ``sklearn``, we have to tell NStack to install some extra operating system libraries inside your module's container. NStack lets us specify these in our ``nstack.yaml``` configuration file in the ``packages`` section. Let's add the following packages:
+
+.. code :: yaml
+
+    packages: ['numpy', 'python3-scikit-learn.x86_64', 'scipy', 'python3-scikit-image.x86_64', 'python3-pandas.x86_64']
```
Comment (jonboulle) [inline advanced_start/more.rst]: it sucks that we have to specify architecture for some of these - is that definitely the case?

File: advanced_start/more.rst [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+
+Now we're ready to build and publish our classifier. Remember, even though we run this command locally, our module gets built and published to your NStack Server.
+
+.. code :: bash
+
+    ~/irisclassify/ $ nstack build
+    Building NStack Container module irisclassify. Please wait. This may take some time.
+    Module irisclassify built successfully. Use `nstack list functions` to see all available functions.
+
+We can now see ``irisclassify.predict`` in the list of existing functions (along with previously built functions like demo.numChars) by running the suggested command nstack list functions
```
Comment (jonboulle) [inline advanced_start/more.rst]: demo.numChars -> `demo.numChars` perhaps

Comment (jonboulle) [reply advanced_start/more.rst]: by running the suggested command

File: advanced_start/more.rst [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+
+.. code :: bash
+ 
+   ~/irisclassify/ $ nstack list functions
+    irisclassify.predict : (Double, Double, Double, Double) -> Text
+    demo.numChars : Text -> Integer
+
+Our classifier is now published, but to use it we need to connect it to an event-source and sink. In the previous tutorial, we used HTTP as a source, and the NStack log as a sink. We can do the same here by starting the following workflow.
```
Comment (jonboulle) [inline advanced_start/more.rst]: an _event source_ and a _sink_

File: advanced_start/index.rst (hunk 7 above)
Comment (jonboulle) [reply advanced_start/index.rst]: Do we really really have to use "productionise"?

File: advanced_start/workflow_power.rst [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+- Expose an HTTP endpoint which takes four ``Double``\s
+- Send these ``Double``\s to our classifier, ``irisclassify``, which will tell us the species of the iris
+- Count the number of characters in the species of the iris
+- Write the result to the log
+
+We could write the following workflow:
+
+.. code :: bash
+   
+  module irsiworkflow {
```
Comment (jonboulle) [inline advanced_start/workflow_power.rst]: irsi -> iris

File: advanced_start/workflow_power.rst [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+    def multipleSteps = sources.http (Double, Double, Double, Double) { http_path = "/irisendpoint" } | irisclassify.predict | demo.numChars | sinks.log : Integer;
+  }
+
+.. note :: ``numChars`` and ``predict`` can be `composed` together because their types -- or schemas -- match. If ``predict`` wasn't configured to output ``Text``, or ``numChars`` wasn't configured to take ``Text`` as input, NStack would not let you build the following workflow.
+
+Partial workflows
+*****************
+
+All of the workflows that we have written so far have been `fully composed`, which means that they contain a source, one or more functions, and a sink. Many times, you want to split up sources, sinks, and functions into separate pieces you can share and reuse. In this case, we say that a workflow is `partially composed`, which just means it does not contain a source, one or more functions, and a sink. These workflows cannot be ``start``\ed by themselves, but can be shared and attached to other sources, sinks, or functions to become `fully composed`.
```
Comment (jonboulle) [inline advanced_start/workflow_power.rst]: Can I create a workflow that's just a source and a sink?

File: advanced_start/index.rst (hunk 7 above)
Comment (lanthias) [reply advanced_start/index.rst]: Good point on the former. On the latter, it's quite a common term but is a bit clunky. Do you prefer operationalise?

File: advanced_start/more.rst (hunk 8 above)
Comment (lanthias) [reply advanced_start/more.rst]: :+1:

File: advanced_start/more.rst (hunk 9 above)
Comment (lanthias) [reply advanced_start/more.rst]: :+1:

File: advanced_start/more.rst (hunk 10 above)
Comment (lanthias) [reply advanced_start/more.rst]: :+1:

File: advanced_start/more.rst (hunk 11 above)
Comment (lanthias) [reply advanced_start/more.rst]: Probably not - can it be left out typically with dnf?

File: advanced_start/more.rst (hunk 12 above)
Comment (lanthias) [reply advanced_start/more.rst]: :+1:

File: advanced_start/more.rst (hunk 13 above)
Comment (lanthias) [reply advanced_start/more.rst]: :+1:

File: advanced_start/workflow_power.rst (hunk 14 above)
Comment (lanthias) [reply advanced_start/workflow_power.rst]: Good catch..

File: advanced_start/workflow_power.rst (hunk 15 above)
Comment (lanthias) [reply advanced_start/workflow_power.rst]: Yes, it's a common theme for data integration.

Comment (jonboulle) [reply advanced_start/workflow_power.rst]: OK, so "one or more functions" should be "zero or more functions"? (but would suggest tweaking the wording)

File: advanced_start/more.rst (hunk 11 above)
Comment (jonboulle) [reply advanced_start/more.rst]: yup

File: advanced_start/workflow_power.rst (hunk 15 above)
Comment (lanthias) [reply advanced_start/workflow_power.rst]: Yes good catch

**Review by jonboulle — APPROVED** [review]
I think this is good for now (please squash)

**Comment by lanthias** [comment]
It's still probably incomplete at then end (it doesn't really have a conclusion), and it stills needs to be changed to use the newest UX syntax on modules and DSL.

File: advanced_start/more.rst [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+If we are trying to find the species based on the sepal and petal measurements, this means these measurements are going to be the input to our classifier module, with text being the output. This means we need to write a function in Python which takes four ``Double``\s and returns ``Text``.
+
+Creating your classifier module
+******************************
+
+To begin, let's make a new directory called ``irisclassify``, ``cd`` into it, and initialise a new module:
+
+.. code:: bash
+    
+    ~/ $ mkdir Irisclassify; cd Irisclassify
```
Comment (rjmk) [inline advanced_start/more.rst]: If we want to follow the directory structure of the examples, this should be `Iris.Classify`

File: advanced_start/more.rst [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+    ~/Irisclassify/ $ nstack init python
+    python module 'Irisclassify' successfully initialised at ~/Irisclassify
+
+Next, let's download our training data so we can use it in our module.
+
+.. code:: bash
+
+    ~/Irisclassify/ $ curl -O https://raw.githubusercontent.com/nstackcom/nstack-examples/master/iris/irisclassify/train.csv
```
Comment (rjmk) [inline advanced_start/more.rst]: curl -O https://raw.githubusercontent.com/nstack/nstack-examples/master/iris/Iris.Classify/train.csv

File: advanced_start/more.rst [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+****************
+
+As we know what the input and output of our classifier is going to look like, let's edit the ``api`` section of ``nstack.yaml`` to define our API (i.e. the entry-point into our module). By default, a new module contains a sample function ``numChars``, which we replace with our definition. We're going to call the function we write in Python ``predict``, which means we fill in the ``api`` section of ``nstack.yaml`` as follows:
+
+.. code :: java
+
+    api : |
+          predict : (Double, Double, Double, Double) -> Text
+
+This means we are exposing a single function ``predict``, which takes a record of four ``Double``\s (the measurements) and returns ``Text`` (the iris species).
```
Comment (rjmk) [inline advanced_start/more.rst]: Would be kind of cool to show off sum types here. How many species are there?

File: advanced_start/more.rst [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+ 
+   ~/irisclassify/ $ nstack list functions
+    Irisclassify.predict : (Double, Double, Double, Double) -> Text
+    Demo.numChars : Text -> Integer
+
+Our classifier is now published, but to use it we need to connect it to an event source and sink. In the previous tutorial, we used HTTP as a source, and the NStack log as a sink. We can do the same here by starting the following workflow.
+
+.. code :: bash
+   
+    ~/Irisclassify/ $ nstack start "sources.http<(Double, Double, Double, Double)> { http_path : "/irisendpoint" } | irisclassify.predict | sinks.log<Text>"
```
Comment (rjmk) [inline advanced_start/more.rst]: This is sadly still not possible. You need a complete workflow for `nstack start` which could be referenced unqualified. The current example has a workflow and a project yaml, which could be recreated?

**Review by rjmk — COMMENTED** [review]
Just looked at the first file so far

File: advanced_start/more.rst (hunk 16 above)
Comment (lanthias) [reply advanced_start/more.rst]: Cool, I'll update everything to use this format.

**Comment by lanthias** [comment]
This is now ready for review to merge.

**Review by mands — APPROVED** [review]

**Pull request merged**